spring-cloud-stream-binder-jms-common/src/main/java/org/springframework/cloud/stream/binder/jms/utils/Base64UrlNamingStrategy.java

```diff
@@ -0,0 +1,39 @@
+package org.springframework.cloud.stream.binder.jms.utils;
+
+import java.nio.ByteBuffer;
+import java.util.UUID;
+
+import org.springframework.util.Base64Utils;
+
+/**
+ * Taken from spring-cloud/spring-cloud-stream-binder-rabbit. See
+ * https://github.com/spring-cloud/spring-cloud-stream-binder-rabbit/blob/master/spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java#L90
+ *
+ * @author Donovan Muller
+ */
+public class Base64UrlNamingStrategy implements AnonymousNamingStrategy {
+
+    private String prefix = "spring.gen-";
+
+    public Base64UrlNamingStrategy() {
+    }
+
+    public Base64UrlNamingStrategy(String prefix) {
+        this.prefix = prefix;
+    }
+
+    @Override
+    public String generateName() {
+        return generateName(this.prefix);
+    }
+
+    @Override
+    public String generateName(String prefix) {
+        UUID uuid = UUID.randomUUID();
+        ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
+        bb.putLong(uuid.getMostSignificantBits()).putLong(uuid.getLeastSignificantBits());
+        // Convert to base64 and remove trailing =
+        return prefix + Base64Utils.encodeToUrlSafeString(bb.array()).replaceAll("=", "")
+                .replaceAll("-", "");
```

Comment:
URL-safe Base64 is standard Base64 but with `+` replaced with `-` and `/` replaced with `_`.

By replacing `"-"` with `""` here, there is an (admittedly tiny) possibility that the name will not be unique.

`...foo-bar...` and `...fo-obar` will become the same value.

We could replace `-` with `$` which is a Java letter...

>An identifier is an unlimited length sequence of Java letters and Java digits, the first of which must be a Java letter. A letter is any character for which the method Character.isJavaLetter returns true. This includes _ and $. A letter or digit is any character for which the method Character.isJavaLetterOrDigit returns true.

We should also prevent a zero-length prefix; otherwise the Base64 could begin with a number, which is not allowed.